test/test_optim.py

```diff
@@ -32,6 +32,10 @@ def drosenbrock(tensor):
 
 
 class TestOptim(TestCase):
+    def setUp(self):
+        super(TestOptim, self).setUp()
+        self.exact_dtype = True
```

Comment:
Don't do this; `exact_dtype` is intended to be a class global property. Just set `exact_dtype = True` directly in the class, like in the other PR.

Reply:
I'm not sure my understanding is correct. Doing `self.exact_dtype = True` in a class member function should be equivalent to directly set `exact_dtype = True` in the class? Also when we modify the inherited attribute exact_dtype, I guess it won't affect the attribute value of the parent's other subclasses, since they instantiate a different class object.

Reply:
No it is not equivalent (although in this particular case I can believe that they are functionally equivalent) 

```
ezyang-mbp:Dev ezyang$ python arf.py
True
Traceback (most recent call last):
  File "arf.py", line 9, in <module>
    print(A2.b)
AttributeError: class A2 has no attribute 'b'
ezyang-mbp:Dev ezyang$ cat arf.py
class A:
    b = True

class A2:
    def __init__(self):
        self.b = True

print(A.b)
print(A2.b)
```